activerecord/lib/active_record/connection_adapters/abstract_mysql_adapter.rb

```diff
@@ -250,8 +258,12 @@ def recreate_database(name, options = {})
       def create_database(name, options = {})
         if options[:collation]
           execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT COLLATE #{quote_table_name(options[:collation])}"
+        elsif options[:charset]
+          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET #{quote_table_name(options[:charset])}"
+        elsif supports_longer_index_key_prefix?
+          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET `utf8mb4`"
         else
-          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET #{quote_table_name(options[:charset] || 'utf8mb4')}"
+          execute "CREATE DATABASE #{quote_table_name(name)} DEFAULT CHARACTER SET `utf8`"
```

Comment:
Perhaps we can safely raise an exception now. If `:charset` is not specified and large prefixes / utf8mb4 are not supported, we don't want to silently default to 3-byte anymore.